integration-test/http/http_system_test.c

```diff
@@ -0,0 +1,369 @@
+    requestInfo.pathLen = strlen( pPath );
+
+    /* Set "Connection" HTTP header to "keep-alive" so that multiple requests
+     * can be sent over the same established TCP connection. */
+    requestInfo.reqFlags = HTTP_REQUEST_KEEP_ALIVE_FLAG;
+
+    /* Set the buffer used for storing request headers. */
+    requestHeaders.pBuffer = userBuffer;
+    requestHeaders.bufferLen = USER_BUFFER_LENGTH;
+
+    httpStatus = HTTPClient_InitializeRequestHeaders( &requestHeaders,
+                                                      &requestInfo );
+    TEST_ASSERT_EQUAL( HTTP_SUCCESS, httpStatus );
+    TEST_ASSERT_NOT_EQUAL( 0, requestHeaders.headersLen );
+
+    /* Initialize the response object. The same buffer used for storing
+     * request headers is reused here. */
+    response.pBuffer = userBuffer;
+    response.bufferLen = USER_BUFFER_LENGTH;
+
+    LogDebug( ( "Sending HTTP %.*s request to %.*s%.*s...",
+                ( int32_t ) requestInfo.methodLen, requestInfo.method,
+                ( int32_t ) SERVER_HOST_LENGTH, SERVER_HOST,
+                ( int32_t ) requestInfo.pathLen, requestInfo.pPath ) );
+    LogDebug( ( "Request Headers:\n%.*s\n"
+                "Request Body:\n%.*s\n",
+                ( int32_t ) requestHeaders.headersLen,
+                ( char * ) requestHeaders.pBuffer,
+                ( int32_t ) REQUEST_BODY_LENGTH, REQUEST_BODY ) );
+
+    /* Send request to HTTP server. */
+    httpStatus = HTTPClient_Send( pTransportInterface,
+                                  &requestHeaders,
+                                  ( uint8_t * ) REQUEST_BODY,
+                                  REQUEST_BODY_LENGTH,
+                                  &response,
+                                  0 );
+
+    /* If a network error is found, retry sending request for a count of MAX_RETRY_COUNT.
+     *  Proceed when MAX_RETRY_COUNT has been hit, or when a successful request is sent,
+     *  whichever comes first. */
+    if( ( httpStatus == HTTP_NETWORK_ERROR ) && ( retryCount < MAX_RETRY_COUNT ) )
+    {
+        LogDebug( ( "A network error has occured, retrying request." ) );
+        resetTest();
+        sendHttpRequest( pTransportInterface, pMethod, pPath, ++retryCount );
+    }
+    else
+    {
+        TEST_ASSERT_EQUAL( HTTP_SUCCESS, httpStatus );
+
+        LogDebug( ( "Received HTTP response from %.*s%.*s...\n"
+                    "Response Headers:\n%.*s\n"
+                    "Response Status:\n%u\n"
+                    "Response Body:\n%.*s\n",
+                    ( int32_t ) SERVER_HOST_LENGTH, SERVER_HOST,
+                    ( int32_t ) requestInfo.pathLen, requestInfo.pPath,
+                    ( int32_t ) response.headersLen, response.pHeaders,
+                    response.statusCode,
+                    ( int32_t ) response.bodyLen, response.pBody ) );
+
+        /* Close connection if a "Connection: close" header is found */
+        if( response.respFlags == HTTP_RESPONSE_CONNECTION_CLOSE_FLAG )
```

Comment:
I think this is redundant because tearDown will close the connection anyways after the test?

Reply:
That's true -- removed it, thanks.